### PR TITLE
SNOW-3258878: SQLCancel scaffolding

### PR DESCRIPTION
**Pull request by sfc-gh-ddas**
### TL;DR

Implements SQLCancel functionality for the ODBC driver with proper state management, cross-thread cancellation support, and comprehensive test coverage.

### What changed?

- **Added SQLCancel API**: Exported `SQLCancel` function in `exports.def` and implemented the C API entry point in `c_api.rs`
- **Enhanced statement state management**: Split the previous `Executed` and `NoResultSet` states into more specific states (`QueryExecuted`, `DdlExecuted`, `DmlExecuted`) to better track statement types and cursor status
- **Implemented cancellation infrastructure**: Added `CancellationToken` to `Statement` struct and wired it through execution paths for future async cancellation support
- **Added cursor state validation**: SQLExecDirect and SQLExecute now return error 24000 (Invalid cursor state) when called with an open cursor, requiring explicit cursor closure
- **Enhanced error handling**: Added new error types `CursorAlreadyOpen`, `DataNotFetched`, and `OperationCanceled` with appropriate SQL state mappings
- **Improved DML handling**: Zero-row DML operations now return `SQL_NO_DATA` instead of opening a cursor
- **Added comprehensive test suite**: 30+ test cases covering basic cancellation, cross-thread scenarios, state preservation, and edge cases

### How to test?

Run the new test suite in `odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp` which covers:
- Basic cancellation scenarios (idle statements, after execution, after fetch)
- Cross-thread cancellation with long-running queries
- State preservation (bindings, attributes, cursor position)
- Error conditions and edge cases
- Platform-specific behavior differences between Windows and Unix ODBC Driver Managers

### Why make this change?

- **ODBC compliance**: SQLCancel is a required ODBC API function that was missing from the driver
- **Application compatibility**: Many ODBC applications expect SQLCancel to be available for query interruption and cursor management
- **Better state management**: The refined state machine provides clearer semantics for different statement types (DDL vs DML vs queries)
- **Foundation for async support**: The cancellation token infrastructure prepares for future asynchronous execution and server-side query cancellation
- **Improved error handling**: Proper cursor state validation prevents application bugs and aligns with ODBC 3.x specifications

### CODE REVIEW

**Comment by sfc-gh-ddas** [comment]
* **#752** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/752?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#769](https://github.com/snowflakedb/universal-driver/pull/769) <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/769?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#857](https://github.com/snowflakedb/universal-driver/pull/857) <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/857?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#751** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/751?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#669** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/669?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/669?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp [hunk 1]
```diff
@@ -336,8 +357,350 @@ TEST_CASE_METHOD(StmtDefaultDSNFixture, "SQLCancel: Diagnostics after cancel err
...
+
+  SQLRETURN cancel_ret = SQLCancel(stmt_handle());
+  REQUIRE(cancel_ret == SQL_SUCCESS);
+
+  int polls = 0;
+  SQLRETURN poll_ret = SQL_STILL_EXECUTING;
+  while (poll_ret == SQL_STILL_EXECUTING && ++polls < kMaxPollIterations) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    poll_ret = SQLExecDirect(stmt_handle(), sqlchar("SELECT COUNT(*) FROM TABLE(GENERATOR(TIMELIMIT => 30))"), SQL_NTS);
+  }
```
[the remark this answers is not in the document]
Comment (sfc-gh-ddas) [reply odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp:690]: This is necessary to ensure the query is in-flight when cancel fires.

File: odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp
Comment (sfc-gh-pbulawa) [inline odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp:1]: Please add tests:

- SQLCancel after error recovery and re-execution. 
- SQLCancel on an allocated-but-never-used statement, then free. Tests cover cancel-on-idle, but don't verify that SQLFreeHandle(SQL_HANDLE_STMT) succeeds after cancel on a fresh statement. This confirms cancel doesn't corrupt the handle state.
- Parameter binding preservation after cancel with open cursor. The existing "Preserves bound parameters" test cancels on a prepared-but-not-executed statement. A variant should cancel after execution with an open cursor, close the cursor, re-execute with the same bindings, and verify values.

**Review by sfc-gh-pbulawa — CHANGES_REQUESTED** [review]

File: odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp
Comment (sfc-gh-ddas) [reply odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp:1]: Added tests for these.

File: odbc_tests/BehaviorDifferences.yaml [hunk 2]
```diff
@@ -312,3 +311,28 @@ behavior_differences:
...
+    description: |
+      OLD driver: SQLCancel unconditionally closes/invalidates the cursor,
+      causing subsequent SQLFetch to return HY010 and SQLExecDirect to return
+      24000. This is a known deviation from the ODBC 3.5 specification.
+      NEW driver: SQLCancel on a synchronous (already-completed) operation is
+      a no-op per the ODBC 3.5 spec. The cursor remains open and usable.
+      Spec reference: https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlcancel-function
+      Impact: Applications that rely on SQLCancel closing cursors will need
+      to use SQLFreeStmt(SQL_CLOSE) or SQLCloseCursor instead.
+  31:
```
Comment (sfc-gh-jszczerbinski) [inline odbc_tests/BehaviorDifferences.yaml:324]: Are you sure about this behavior difference? I think you are applying the documentation incorrectly here. It says we need call close cursor before reexecuting, but in the tests you are applying this BD where there is no cursor open (for example INSERT returns no result set -> therefore doesn't open cursor)

File: odbc_tests/BehaviorDifferences.yaml [hunk 3]
```diff
@@ -312,3 +311,28 @@ behavior_differences:
       field incorrect), caught locally by the driver.
       NEW driver: Sends the query to the server, which returns SQL_ERROR with
       SQLSTATE 42000 (Syntax error: Bind variable ? not set).
+  30:
```
Comment (sfc-gh-jszczerbinski) [inline odbc_tests/BehaviorDifferences.yaml:315]: Looks like DM behavior difference between Linux and Windows. Good catch

File: odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp [hunk 4]
```diff
@@ -127,30 +133,38 @@ TEST_CASE_METHOD(StmtDefaultDSNFixture, "SQLCancel: After cancel on executed pre
...
-
   SQLRETURN ret = SQLExecDirect(stmt_handle(), sqlchar("SELECT 1"), SQL_NTS);
   REQUIRE(ret == SQL_SUCCESS);
 
   ret = SQLCancel(stmt_handle());
   REQUIRE(ret == SQL_SUCCESS);
 
+  // Both drivers return 24000 on re-execution after cancel, but for
+  // different reasons:
+  // - Old driver (BD#30): cancel closes the cursor on Unix, but leaves the
```
Comment (sfc-gh-jszczerbinski) [inline odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp]: Can we disambiguate here somehow in code?

File: odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp [hunk 5]
```diff
@@ -164,37 +178,99 @@ TEST_CASE_METHOD(StmtDefaultDSNFixture, "SQLCancel: Statement recoverable via SQ
...
                  "[odbc-api][cancel][terminating_statement]") {
-  SKIP_NEW_DRIVER_NOT_IMPLEMENTED();
-
   SQLRETURN ret = SQLExecDirect(stmt_handle(), sqlchar("SELECT 1"), SQL_NTS);
   REQUIRE(ret == SQL_SUCCESS);
 
   ret = SQLCancel(stmt_handle());
   REQUIRE(ret == SQL_SUCCESS);
 
   WINDOWS_ONLY {
```
Comment (sfc-gh-jszczerbinski) [inline odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp:210]: Nit: in general could swap PLATFORM_ONLY statements be inside DRIVER_ONLY? So we can highlight that new driver behave the same way on both oses and old driver doesn't? Similiar for other tests like that

File: odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp [hunk 6]
```diff
@@ -307,27 +423,36 @@ TEST_CASE_METHOD(StmtDefaultDSNFixture, "SQLCancel: Cancels data-at-execution op
...
-// ============================================================================
-
-TEST_CASE_METHOD(StmtDefaultDSNFixture, "SQLCancel: Diagnostics after cancel error state",
+TEST_CASE_METHOD(StmtDefaultDSNFixture, "SQLCancel: Re-execute immediately after canceling data-at-execution",
                  "[odbc-api][cancel][terminating_statement]") {
   SKIP_NEW_DRIVER_NOT_IMPLEMENTED();
 
-  SQLRETURN ret = SQLExecDirect(stmt_handle(), sqlchar("SELECT 1"), SQL_NTS);
+  SQLRETURN ret = SQLPrepare(stmt_handle(), sqlchar("SELECT ?"), SQL_NTS);
   REQUIRE(ret == SQL_SUCCESS);
```
Comment (sfc-gh-jszczerbinski) [inline odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp]: You might want to migrate to matchers in those tests: REQUIRE_THAT(OdbcResult(stmt, ret), IsSuccess) gives much better debug info

File: odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp [hunk 7]
```diff
@@ -336,8 +461,350 @@ TEST_CASE_METHOD(StmtDefaultDSNFixture, "SQLCancel: Diagnostics after cancel err
...
+}
+
+// ============================================================================
+// SQLCancel - Async Cancel
+// ============================================================================
+
+TEST_CASE_METHOD(StmtDefaultDSNFixture, "SQLCancel: Async cancel interrupts execution with HY008",
+                 "[odbc-api][cancel][terminating_statement][async]") {
+  SKIP_NEW_DRIVER_NOT_IMPLEMENTED();
+  SKIP_OLD_DRIVER("BD#30", "Reference driver async cancel does not interrupt in-progress operations");
```
Comment (sfc-gh-jszczerbinski) [inline odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp]: Is this the same behavior difference that is descibed in YAML?

File: odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp [hunk 8]
```diff
@@ -336,8 +461,350 @@ TEST_CASE_METHOD(StmtDefaultDSNFixture, "SQLCancel: Diagnostics after cancel err
...
+  // The HY008 from the canceled SQLExecDirect should be the only record.
+  auto records = get_diag_rec(SQL_HANDLE_STMT, stmt);
+  REQUIRE(records.size() == 1);
+  REQUIRE(records[0].sqlState == "HY008");
+}
+
+TEST_CASE_METHOD(StmtDefaultDSNFixture, "SQLCancel: HY018 when server declines cancel request",
+                 "[odbc-api][cancel][terminating_statement][cross_thread]") {
+  SKIP_NEW_DRIVER_NOT_IMPLEMENTED();
+  SKIP_OLD_DRIVER("BD#30", "Snowflake server does not decline cancel requests");
```
Comment (sfc-gh-jszczerbinski) [inline odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp]: Is this the same behavior difference?

**Review by sfc-gh-jszczerbinski — CHANGES_REQUESTED** [review]
In general looks good, two main points:  
\- I disagree with some of the behavior differences  
\- Let's discuss the overall approach for cancellation, I don't think this extra atomic field is the right choice here, but we can tackle that in the next PR

File: odbc_tests/BehaviorDifferences.yaml (hunk 2 above)
Comment (sfc-gh-ddas) [reply odbc_tests/BehaviorDifferences.yaml:324]: Added new state for `Statement` and renamed existing states to make it more clear.

File: odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp (hunk 7 above)
Comment (sfc-gh-ddas) [reply odbc_tests/tests/odbc-api/terminating_statement/cancel_tests.cpp]: This is a different BD, Old driver doesn't seem to do the Async cancel correctly.

File: odbc/src/api/statement.rs [hunk 9]
```diff
@@ -22,8 +22,60 @@ use sf_core::protobuf::generated::database_driver_v1::{
...
+use tokio_util::sync::CancellationToken;
 use tracing;
 
+// ---------------------------------------------------------------------------
+// Cancel registry: stores CancellationTokens keyed by statement handle address.
+// SQLCancel looks up the token here without dereferencing the Statement,
+// eliminating aliasing UB when called from another thread.
+// ---------------------------------------------------------------------------
+
+static CANCEL_REGISTRY: LazyLock<Mutex<HashMap<usize, CancellationToken>>> =
```
Comment (sfc-gh-jszczerbinski) [inline odbc/src/api/statement.rs]: We shouldn't rely on global state in ODBC, as Driver manager can load and unload the binary and statics do not get dropped in Rust (see how we handle [runtime.rs](http://runtime.rs))

Also we have a global mutex here that could receive a lot of contention, instead can we store CancelationToken as RwLock\<Arc\<CancellationToken>> in stmt/conn/env structs?

SQLCancel would take current cancellation token, call cancel on it and swap it with a new token for subsequent operations to use?

Other operations would read the cancelation token at the beginning of the operation and await cancels where needed.

**Review by sfc-gh-jszczerbinski — CHANGES_REQUESTED** [review]
Please avoid globals in ODBC

File: odbc/src/api/statement.rs [hunk 10]
```diff
@@ -38,6 +90,18 @@ fn exec_direct_impl(statement_handle: sql::Handle, statement_text: &str) -> Odbc
...
+        stmt.state.as_ref(),
+        StatementState::QueryExecuted { .. }
+            | StatementState::Fetching { .. }
+            | StatementState::Done { .. }
+    ) {
+        tracing::error!("exec_direct: cursor is already open");
+        return CursorAlreadyOpenSnafu.fail();
+    }
+
+    let _cancel_token = reset_cancel_token(statement_handle as usize);
```
Comment (sfc-gh-pbulawa) [inline odbc/src/api/statement.rs]: If the connection is `Disconnected`, a new token is allocated in the registry (replacing the existing one) and then immediately discarded when the function returns an error. Let's move it inside the `Connected` branch.

File: odbc/src/api/statement.rs [hunk 11]
```diff
@@ -197,17 +275,20 @@ pub fn execute(statement_handle: sql::Handle) -> OdbcResult<()> {
...
+        StatementState::QueryExecuted { .. }
             | StatementState::Fetching { .. }
             | StatementState::Done { .. }
     ) {
         tracing::error!("execute: cursor is already open");
-        return InvalidCursorStateSnafu.fail();
+        return CursorAlreadyOpenSnafu.fail();
     }
 
+    let _cancel_token = reset_cancel_token(statement_handle as usize);
```
Comment (sfc-gh-pbulawa) [inline odbc/src/api/statement.rs]: If the connection is `Disconnected`, a new token is allocated in the registry (replacing the existing one) and then immediately discarded when the function returns an error. Let's move it inside the `Connected` branch.

**Review by sfc-gh-pbulawa — CHANGES_REQUESTED** [review]

File: odbc/src/api/statement.rs (hunk 9 above)
Comment (sfc-gh-ddas) [reply odbc/src/api/statement.rs]: When `SQLExecDirect` runs, it dereferences the statement handle to get an exclusive mutable reference to the entire struct. If `SQLCancel` is called from another thread while `SQLExecDirect` is in-flight, and `cancel()` also accesses a field on `Statement` we'd have two concurrent references to the same struct. I see the downsides of the global state, however the only way we can still have the cancel token on `Statement` is by using a pointer to it and using `addr_of!` to get to the cancel token and create a `&RwLock<>` scoped to just that field. This is very hacky imo, what do you think?

Comment (sfc-gh-jszczerbinski) [reply odbc/src/api/statement.rs]: Statements not being thread safe is separate issue and we should deal with it in time, it shouldn't stop us from having a single field that is thread safe. This solution is not acceptable to me

**Review by sfc-gh-jszczerbinski — CHANGES_REQUESTED** [review]

File: odbc/src/api/statement.rs (hunk 9 above)
Comment (sfc-gh-jszczerbinski) [reply odbc/src/api/statement.rs]: > however the only way we can still have the cancel token on `Statement` is by using a pointer to it and using `addr_of!`

Why is that?

Comment (sfc-gh-ddas) [reply odbc/src/api/statement.rs]: Removed the global registry and moved the `CancellationToken` in the `Statement` struct. `cancel()` now accesses it via `stmt_from_handle()`. The pre-existing aliasing issue from `stmt_from_handle` creating `&mut Statement` concurrently will be addressed by the future handle manager

**Review by sfc-gh-jszczerbinski — APPROVED** [review]

**Review by sfc-gh-pbulawa — APPROVED** [review]
